### PR TITLE
Python3.10 v2

### PR DESCRIPTION
**Pull request by esc**
This is a rebuild of the PR at #7381 -- using what we learnt in #7544 and #7529

### CODE REVIEW

File: numba/tests/test_analysis.py [hunk 1]
```diff
@@ -246,8 +246,6 @@ def impl(x):
             x_is_none_work = False
             if x is None:
                 x_is_none_work = True
-            else:
-                pass  # force the True branch exit to be on backbone
```
Comment (stuartarchibald) [inline numba/tests/test_analysis.py:249]: I'm not sure that this change is valid. Am looking into it.

Comment (esc) [reply numba/tests/test_analysis.py:249]: it originates from `https://github.com/numba/numba/pull/7381/commits/036d43b778a8dd16791ee4b8cc893c5bdc2f5883

**Comment by esc** [comment]
The errors:

```
======================================================================
ERROR: test_floats (numba.tests.test_hashing.TestNumberHashing)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/runner/work/1/s/numba/tests/test_hashing.py", line 161, in test_floats
    self.check_floats(np.float32)
  File "/Users/runner/work/1/s/numba/tests/test_hashing.py", line 151, in check_floats
    for a in self.float_samples(typ):
  File "/Users/runner/work/1/s/numba/tests/test_hashing.py", line 125, in float_samples
    a.append(typ(float('nan')))
AttributeError: 'numpy.ndarray' object has no attribute 'append'

======================================================================
ERROR: test_complex (numba.tests.test_hashing.TestNumberHashing)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/runner/work/1/s/numba/tests/test_hashing.py", line 165, in test_complex
    self.check_complex(np.complex64, np.float32)
  File "/Users/runner/work/1/s/numba/tests/test_hashing.py", line 156, in check_complex
    for a in self.complex_samples(typ, float_ty):
  File "/Users/runner/work/1/s/numba/tests/test_hashing.py", line 131, in complex_samples
    for imag in self.float_samples(float_ty):
  File "/Users/runner/work/1/s/numba/tests/test_hashing.py", line 125, in float_samples
    a.append(typ(float('nan')))
AttributeError: 'numpy.ndarray' object has no attribute 'append'

----------------------------------------------------------------------
Ran 353 tests in 88.531s

FAILED (errors=2, skipped=2, expected failures=1)
```

Are likely related to: "the hashing patch" currently at 40de856

**Comment by esc** [comment]
> The errors:
> 
> ```
> ======================================================================
> ERROR: test_floats (numba.tests.test_hashing.TestNumberHashing)
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/Users/runner/work/1/s/numba/tests/test_hashing.py", line 161, in test_floats
>     self.check_floats(np.float32)
>   File "/Users/runner/work/1/s/numba/tests/test_hashing.py", line 151, in check_floats
>     for a in self.float_samples(typ):
>   File "/Users/runner/work/1/s/numba/tests/test_hashing.py", line 125, in float_samples
>     a.append(typ(float('nan')))
> AttributeError: 'numpy.ndarray' object has no attribute 'append'
> 
> ======================================================================
> ERROR: test_complex (numba.tests.test_hashing.TestNumberHashing)
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/Users/runner/work/1/s/numba/tests/test_hashing.py", line 165, in test_complex
>     self.check_complex(np.complex64, np.float32)
>   File "/Users/runner/work/1/s/numba/tests/test_hashing.py", line 156, in check_complex
>     for a in self.complex_samples(typ, float_ty):
>   File "/Users/runner/work/1/s/numba/tests/test_hashing.py", line 131, in complex_samples
>     for imag in self.float_samples(float_ty):
>   File "/Users/runner/work/1/s/numba/tests/test_hashing.py", line 125, in float_samples
>     a.append(typ(float('nan')))
> AttributeError: 'numpy.ndarray' object has no attribute 'append'
> 
> ----------------------------------------------------------------------
> Ran 353 tests in 88.531s
> 
> FAILED (errors=2, skipped=2, expected failures=1)
> ```
> 
> Are likely related to: "the hashing patch" currently at [40de856](https://github.com/numba/numba/commit/40de856b9754852ff5eec84a8ca0fc4a58646612)

@stuartarchibald would suggest the following fix for this patch:

```
diff --git i/numba/tests/test_hashing.py w/numba/tests/test_hashing.py
index fbac744045..08443bb52e 100644
--- i/numba/tests/test_hashing.py
+++ w/numba/tests/test_hashing.py
@@ -117,14 +117,14 @@ class BaseTest(TestCase):
                 yield a + a.mean()

         # Infs, nans, zeros, magic -1
-        a = typ([0.0, 0.5, -0.0, -1.0, float('inf'), -float('inf'),])
+        a = [0.0, 0.5, -0.0, -1.0, float('inf'), -float('inf'),]

         # Python 3.10 has a hash for nan based on the pointer to the PyObject
         # containing the nan, skip this input and use explicit test instead.
         if utils.PYVERSION < (3, 10):
-            a.append(typ(float('nan')))
+            a.append(float('nan'))

-        yield a
+        yield typ(a)

     def complex_samples(self, typ, float_ty):
         for real in self.float_samples(float_ty):
```

File: numba/core/untyped_passes.py [hunk 2]
```diff
@@ -946,9 +946,50 @@ def foo():
         """
         elif_tplt = "\n\telif PLACEHOLDER_INDEX in (%s,):\n\t\tSENTINEL = None"
 
+        # Note regarding the insertion of the garbage/defeat variables below:
+        # These values have been designed and inserted to defeat a specific
+        # beahviour of the cpython optimizer. The optimization was introduced
```
Comment (sklam) [inline numba/core/untyped_passes.py:951]:
```suggestion
        # behaviour of the cpython optimizer. The optimization was introduced
```

File: numba/cpython/hashing.py [hunk 3]
```diff
@@ -76,7 +82,15 @@ def _Py_HashDouble(v):
             else:
                 return -_PyHASH_INF
         else:
-            return _PyHASH_NAN
+            # Python 3.10 does not use `_PyHASH_NAN`.
+            # https://github.com/python/cpython/blob/2c4792264f9218692a1bd87398a60591f756b171/Python/pyhash.c#L102   # noqa: E501
+            # Numba returns a pseudo-random number to reflect the spirit of the
+            # change.
+            if _py310_or_later:
+                x = _prng_random_hash()
```
Comment (sklam) [inline numba/cpython/hashing.py:90]: I checked the performance difference here. With the py310 change, it takes about ~9ns. Before, it takes about ~3ns. I was worried that we might be looking a >10x difference. The 2-3x difference for a rare case is acceptable IMO.

File: numba/tests/test_tuples.py [hunk 4]
```diff
@@ -745,8 +745,10 @@ def foo():
 
         with self.assertRaises(errors.UnsupportedError) as raises:
             foo()
-
-        msg = "op_LIST_EXTEND at the start of a block"
+        if utils.PYVERSION == (3, 9):
+            msg = "op_LIST_EXTEND at the start of a block"
+        elif utils.PYVERSION == (3, 10):
+            msg = "SPECIALSTRING"
```
Comment (sklam) [inline numba/tests/test_tuples.py:751]: We should change this to something more informative.

File: numba/core/interpreter.py [hunk 5]
```diff
@@ -1818,6 +1818,8 @@ def op_LIST_EXTEND(self, inst, target, value, extendvar, res):
                 else:
                     ok = False
                     break
+        if ok and build_empty_list is None:
+            raise errors.UnsupportedError("SPECIALSTRING")
```
Comment (sklam) [inline numba/core/interpreter.py:1822]: Change the err msg to something more informative

Comment (esc) [reply numba/core/interpreter.py:1822]: done in 8c7a69d804b9213e2a46c28eccede4eb930d437b

File: numba/tests/test_tuples.py (hunk 4 above)
Comment (esc) [reply numba/tests/test_tuples.py:751]: done in 8c7a69d804b9213e2a46c28eccede4eb930d437b

File: numba/tests/test_analysis.py (hunk 1 above)
Comment (stuartarchibald) [reply numba/tests/test_analysis.py:249]: Try this:
```diff
diff --git a/numba/tests/test_analysis.py b/numba/tests/test_analysis.py
index f89042f..3a75019 100644
--- a/numba/tests/test_analysis.py
+++ b/numba/tests/test_analysis.py
@@ -246,6 +246,8 @@ class TestBranchPrune(TestBranchPruneBase, SerialMixin):
             x_is_none_work = False
             if x is None:
                 x_is_none_work = True
+            else:
+                pass
 
             if x_is_none_work:
                 y = 10
@@ -253,7 +255,14 @@ class TestBranchPrune(TestBranchPruneBase, SerialMixin):
                 y = -3
             return y
 
-        self.assert_prune(impl, (types.NoneType('none'),), [None, None], None)
+        if utils.PYVERSION >= (3, 10):
+            # Python 3.10 creates a block with a NOP in it for the `pass` which
+            # means it gets pruned.
+            self.assert_prune(impl, (types.NoneType('none'),), [False, None],
+                              None)
+        else:
+            self.assert_prune(impl, (types.NoneType('none'),), [None, None],
+                              None)
         self.assert_prune(impl, (types.IntegerLiteral(10),), [True, None], 10)
 ```

Comment (stuartarchibald) [reply numba/tests/test_analysis.py:249]: I don't think there's a way to create the scenario that was originally being tested in Python 3.10, but there's value in having the test for `< 3.10`. In Python 3.10 the `else: pass` ends up as a block with a `NOP` opposed to a direct jump from the `if` to the backbone, the above adjusts the checking for the expected behaviours.

**Comment by esc** [comment]
Note to self: need to double check if public CI is configured for 3.10 as part of this PR

File: numba/tests/test_analysis.py (hunk 1 above)
Comment (esc) [reply numba/tests/test_analysis.py:249]: applied as 3edd2d79f51c68c99e3dae5fe470d54473cdcd3f

**Comment by esc** [comment]
I have rebased this onto `master` and included the patch from #7563 to test 3.10 on public CI via cherry-pick 🍒

File: numba/core/untyped_passes.py [hunk 6]
```diff
@@ -946,9 +946,50 @@ def foo():
...
+        #  File "/numba/numba/core/untyped_passes.py", line 830, \
+        #       in inject_loop_body
+        #   sentinel_exits.add(blk.body[-1].target)
+        # AttributeError: Failed in nopython mode pipeline \
+        #       (step: handles literal_unroll)
+        # Failed in literal_unroll_subpipeline mode pipeline \
+        #       (step: performs mixed container unroll)
+        # 'Return' object has no attribute 'target'
+        #
+        # Which indicates that a Return has been found instead of a Jump
```
Comment (stuartarchibald) [inline numba/core/untyped_passes.py:984]: I'm wondering if there ought to be a "bytecode notes" section in the dev docs that documents roughly all the changes that had to happen to accommodate a given new Python version. A version of this text really belongs in there?

File: numba/cpython/hashing.py (hunk 3 above)
Comment (stuartarchibald) [reply numba/cpython/hashing.py:90]: Thanks for checking @sklam

File: numba/tests/test_analysis.py (hunk 1 above)
Comment (stuartarchibald) [reply numba/tests/test_analysis.py:249]: Is this patch (3edd2d7) on this branch?

File: numba/tests/test_tuples.py (hunk 4 above)
Comment (stuartarchibald) [reply numba/tests/test_tuples.py:751]: Is 8c7a69d on this branch?

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the patch. Given I wrote some of this I cannot review it all, but I've commented on the bits I can. There's a few minor things to look at but it otherwise looks good.

**Comment by stuartarchibald** [comment]
Other things that need updating:
* The README: https://github.com/numba/numba/blame/02f02a59cf7ab3cffbb459ae27f9720c6310d3cb/README.rst#L51
* The classifiers: https://github.com/numba/numba/blob/02f02a59cf7ab3cffbb459ae27f9720c6310d3cb/setup.py#L379

File: numba/tests/test_analysis.py (hunk 1 above)
Comment (esc) [reply numba/tests/test_analysis.py]: now as a5497934cc3faa8d46e12b97d59f166faf5e4161

Comment (esc) [reply numba/tests/test_analysis.py]: now as d7558eb88bc01428c63679124f775ca02a3a593e

File: numba/tests/test_tuples.py (hunk 4 above)
Comment (esc) [reply numba/tests/test_tuples.py]: now as a5497934cc3faa8d46e12b97d59f166faf5e4161

File: numba/core/interpreter.py (hunk 5 above)
Comment (esc) [reply numba/core/interpreter.py]: now as a5497934cc3faa8d46e12b97d59f166faf5e4161

File: numba/core/untyped_passes.py (hunk 6 above)
Comment (esc) [reply numba/core/untyped_passes.py:984]: probably, yes, no one will ever find it again when it is here. Maybe as part of the developer docs: "Notes on Bytecode"?

File: numba/core/untyped_passes.py [hunk 7]
```diff
@@ -946,9 +946,50 @@ def foo():
...
+
         b = ('def foo():\n\tif PLACEHOLDER_INDEX in (%s,):\n\t\t'
              'SENTINEL = None\n%s\n\telse:\n\t\t'
-             'raise RuntimeError("Unreachable")')
+             'raise RuntimeError("Unreachable")\n\t'
+             'temp_defeat_garbage = None\n\t'     # defeat optimizer
+             'obvious_defeat_pattern = None\n\t'  # defeat optimizer
+             'return None\n\t'                    # defeat optimizer
```
[the remark this answers is not in the document]
Comment (esc) [reply numba/core/untyped_passes.py]: Agreed, done in 40d1485ad05776052b9f103ce7f6b33a2992a38a

File: numba/core/untyped_passes.py (hunk 6 above)
Comment (stuartarchibald) [reply numba/core/untyped_passes.py:984]: Assuming this is a new page in the docs, then yes, I think that makes sense. Perhaps it ought to be "Notes on Python version upgrades", so we can add things like what we have to do to deal with the bytecode, but also record other changes that impact Numba, for example the change in `hash(NaN)`, the bytecode offsets change, the cProfiler change needed?

**Comment by esc** [comment]
> Other things that need updating:
> 
> * The README: https://github.com/numba/numba/blame/02f02a59cf7ab3cffbb459ae27f9720c6310d3cb/README.rst#L51
> * The classifiers: https://github.com/numba/numba/blob/02f02a59cf7ab3cffbb459ae27f9720c6310d3cb/setup.py#L379

Done in 2109064da3d966697c56bfd8554387ff4d89a71e

File: numba/core/untyped_passes.py (hunk 6 above)
Comment (esc) [reply numba/core/untyped_passes.py:984]: 👍

**Comment by esc** [comment]
Also, I know where the missing commits went. They were on the `Python3.10` branch and not the `Python3.10_v2` branch... my bad.

File: numba/tests/test_tuples.py [hunk 8]
```diff
@@ -745,8 +745,8 @@ def foo():
 
         with self.assertRaises(errors.UnsupportedError) as raises:
             foo()
-
-        msg = "op_LIST_EXTEND at the start of a block"
+        if utils.PYVERSION >= (3, 9):
+            msg = "op_LIST_EXTEND at the start of a block"
         self.assertIn(msg, str(raises.exception))
```
Comment (sklam) [inline numba/tests/test_tuples.py:750]: The test is already skipped on py<3.9. The `if` is unnecessary. Also, `msg` is only defined in the `if` block. but used after the `if`.

Comment (esc) [reply numba/tests/test_tuples.py]: done in 9b0e74f86eb8e63c966f362a130607f3cc97e612

**Review by sklam — APPROVED** [review]
Looks like everything is addressed. I think this is just pending CI

**Comment by esc** [comment]
> Looks like everything is addressed. I think this is just pending CI

Greaaaat! @sklam this contains code for running on 3.10 on public CI, I would suggest to leave the hack in, so that it can run on PRs also for now. We will need to upgrade the public CI config again anyway, when the Numpy 3.10 packages that we need arrive. OK?

**Comment by esc** [comment]
@stuartarchibald thank you for the patches @sklam thank you for the review!

**Comment by esc** [comment]
@stuartarchibald thank you for the patches @sklam thank you for the review!

**Pull request merged**